documentation/cadl-structure-guidelines.md

```diff
@@ -0,0 +1,127 @@
+# Repository Guidelines for Cadl Specs
+
+## Purpose
+
+We need to formulate a strategy for checking in actual Cadl service specs for Azure. Service teams need to know where
+and how to check these in and tooling needs to know how to find and consume them. While OpenAPI specs traditionally have been just spec
+documents, Cadl "specs" could include service-specific functions, decorators and so forth that elevate them to the level of a
+library, and to maximize the benefits of Cadl, this approach needs to be supported as a first-class citizen.
+
+## Repository
+
+Cadl can co-exist with Swagger within the existing `azure-rest-api-specs` repository. This approach will make it easier to generate Swagger artifacts without needing to sync between repos at the downside of having to live with any baggage associated with the repo as it ages. There are currently over 1,000 issues and 500 open PRs in this repo.
+
+## Structure Overview
+
+This proposal strives to align with [Azure SDK Repo Structure Guidelines](https://azure.github.io/azure-sdk/policies_repostructure.html).
+
+We eliminate the distinction between data-plane and resource-manager, treating all packages as siblings under the service family folder. This is what the SDKs do today. The legacy `data-plane` and `resource-manager` folders where Swagger specs live will remain.
+
+A given Cadl folder could represent any of the following scenarios:
+
+- an SDK and service spec (simplest, common for small services)
+- a service spec only
+- an SDK spec that aggregates many small service specs (one SDK to many microservices)
+- an SDK spec that extracts a portion of a large service spec (many SDKs to a single large service)
+- a shared library that represents neither a service spec nor an SDK
+
+Examples below:
+
+```
+-> specification
+   -> confidentialledger
+      -> ConfidentialLedger               (data-plane SDK + service)
+      -> data-plane                       (swagger)
+      -> resource-manager                 (swagger)
+   -> compute
+      -> Compute.Management               (mgmt SDK + service)
```

Comment:
How about adding a root folder say 'CADL' to keep all packages, Compute.Management, Compute.Management.Shared, etc, so that validation tools running in PR pipeline can be easier to filter out this root folder 'CADL' and no need to worry about the dynamically changed folder names?

Reply:
As we chatted we wanted to keep "cadl" out of the folder structure and try to keep this as simple as possible for the 80% of cases where there will only one mgmt and/or one dpg project folder under the service folder.

Reply:
in a word, although under the same <rp-name> folder, cadl specs and swagger specs will be in separate sibling folders. while swagger specs are under folders with fixed names as 'data-plane' and 'resource-manager', cadl specs are under folders named after SDK package.  However, for a first-time onboarding resource provider, the SDK package name need to be reviewed and decided in the SDK ArchBoard Review, which happens after API spec is frozen for SDK release.  Did you see the contradiction here? In the case, the RP can only randomly name the folder first, and change the folder name later.    On the other hand, because the folder name and number of folders are not fixed, tooling needs to scan all folders under <rp-name> folder to decide what are cadl folders for which SDK type.  It is doable, but not straightforward enough.

Reply:
Beside api specs, examples are also key artifacts.  Not sure how service should define examples when using cadl for API definition?  And what are folders for examples?

Reply:
The folder names don't necessarily map to the exact name of the sdk a lot of times it more closely maps to the resource provider namespace so I don't think that should really be any more of an issue then we had in the past. Also, we should be careful to avoid hardcoding folder locations so that we can rename them if needed later. By the time we do our first PR merge we should have enough confidence in the name that it shouldn't really be an issue.

Reply:
Is there any requirement for data plane(cadl) folder name? such as case convention (pascal case or camal case), namespace convention (does it must include dot separated segments), how many segments of namespace allowed (a.b or a.b.c), etc.  we need to have them documented. Otherwise, user could have different understanding on the rule which will make them inconsistent.
confidentialledger->ConfidentialLedger: two same name with different case is a little confused to me, why should I have two almost duplicated name in paths.

Reply:
I suspect we should use pascal casing with `.` separators very muchs similar to the RP and/or namespace guideline we already have. No real change there.